sdk/core/src/main/java/com/klaviyo/coresdk/networking/NetworkMonitor.kt

```diff
@@ -0,0 +1,111 @@
+package com.klaviyo.coresdk.networking
+
+import android.net.ConnectivityManager
+import android.net.LinkProperties
+import android.net.Network
+import android.net.NetworkCapabilities
+import android.net.NetworkRequest
+import androidx.core.content.ContextCompat
+import com.klaviyo.coresdk.Registry
+
+typealias NetworkObserver = (isConnected: Boolean) -> Unit
+
+/**
+ * Provides methods to react to changes in the application environment
+ */
+interface NetworkMonitor {
+
+    /**
+     * Register an observer to be notified when network connectivity has changed
+     *
+     * @param observer
+     */
+    fun whenNetworkChanged(observer: NetworkObserver)
+
+    /**
+     * Instant check of network connectivity
+     *
+     * @return Boolean
+     */
+    fun isNetworkConnected(): Boolean
+}
+
+/**
+ * Service for monitoring the application lifecycle and network connectivity
+ *
+ * Maintains a list of subscribed observers and notified them all
+ * whenever any internet connectivity change is detected
+ */
+internal object KlaviyoNetworkMonitor : NetworkMonitor {
+
+    private lateinit var networkRequest: NetworkRequest
+
+    private val connectivityManager: ConnectivityManager
+        get() = ContextCompat.getSystemService(
+            Registry.config.applicationContext, ConnectivityManager::class.java
+        ) as ConnectivityManager
```

Comment:
can do just do `getSystemService<ConnectivityManager>()` since we support at least 23, ContextCompat isnt necessary (impl of it checks for >= 23)